### DISCUSSION TITLE
The injected instance into a test-class Constructor is not generated each time the Constructor is ran - this means some state can be shared, unfortunately

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#2455] [Question] [answered]
👋🏻 G'Day Team!

I have some integration tests and I'm following the standard `IClassFixture<CustomWebApplicationFactory>` pattern. Now I thought that [every time a test is setup and ran, a new instance is created](https://xunit.net/docs/shared-context). So yes, a new instance of the _test class_ occurs for each test. ✅

BUT, the constructor accepts a `CustomWebApplicationFactory` ... which _I thought/assumed_ would be a new instance _also_, each time the constructor is ran/generated?. I'm _guessing_ it's like a "Singleton" that is generated once _per test run_ ?

for example, here's the pretty stock standard code/pattern:

```
public class UrlStateTests : IClassFixture<CustomWebApplicationFactory>
{
    private readonly CustomWebApplicationFactory _factory;

    public UrlStateTests(CustomWebApplicationFactoryfactory)
    {
        _factory = factory ?? throw new ArgumentNullException(nameof(factory));
    }
}
```

so with that example, the injected instance (into the ctor) is only created once and reused.

So, is there a way to make sure that if i need a totally new instance of the injected factory, I can ask for that?

An example of this, is different tests wish to define different outcomes when 'ConfiguringServices'. For example, Test1: this might do an HTTP request (urgh .. ) to some api and that fragile test now passes. While Test2 might want to mock the HTTP request always fails, so lets see what happens? Now, this can easily be discussed down to a unit test instead of an integration test .. but lets ignore that right now, etc.

**Comment by bartelink** [comment, marked as answer]
In V2, xUnit's ctor args are not intended to be a generic DI/container thing (see Brad's twitter and/or open issues for ideas about extending that in V3)

The entire point of ClassFixture is for it to be reused - the mechanism exists within xUnit for that to be managed cleanly including cleanup.

If you need a fresh instance per test, you simply can do that yourself using the existing things like Dispose/DisposeAsync etc

**Reply by PureKrome** [reply]
Thanks @bartelink for the really prompt reply - appreciate it!

I sorta don't understand this bit:
> If you need a fresh instance per test, you simply can do that yourself using the existing things like Dispose/DisposeAsync etc

I can't see how the Dispose/DisposeAsync will destroy the `CustomWebApplicationFactoryfactory` instance and generate a new one?

**Reply by bartelink** [reply]
I'm saying you'd do it manually: construct and assign it to a field in the ctor, and if it needs Disposal, you can `field.?Dispose()` in a Dispose override

**Reply by bartelink** [reply]
While it could be argued that a FreshClassFixture concept and/or having that be done for all ctor args that are not class or collection fixtures would eb a useful addition to xUnit v2, it also makes for more confusion and/or assumptions, e.g. if you forget to put in the marker interface or you put the wrong one, you could be a long time with wrong assumptions

**Reply by PureKrome** [reply]
Ah ok .. so don't use the `: IClassFixture<CustomWebApplicationFactory>` then? just `new CustomWebApplicationFactory()` in the ctor?

**Reply by bartelink** [reply]
Yes. In V2 only collection fixture items and declared IClassFixtures can/should be accessed via the ctor.

**Reply by PureKrome** [reply]
Totally works! So what I misunderstood was this (emphasis mine):

> Class Fixtures
When to use: when you want to create a single test context and **share it among all the tests in the class**, and have it cleaned up after all the tests in the class have finished.

So the `IClassFixture<T>` is doing _exactly_ what it's designed to do! I just misread. All G!

thanks heaps @bartelink for the awesome help - really really appreciate it.

![](https://media.giphy.com/media/RipfZWzjUDH25euMpM/giphy.gif)

**Reply by bartelink** [reply]
yw; I (and others) have spent inordinate amounts of time reading that article as there are definitely plenty subtleties to it all

In fairness it does (like all things in xUnit.NET) end up bullet proof and minimal in the end once you grok it, with no special cases to reason about or 20 ways to do the same thing.

I'm looking forward to seeing where the V3 design lands on - there's a lot of ways of skinning the cat, but there's a history of good taste/design winning out https://twitter.com/bradwilson/status/1476636947707695106
